File: DoublyLinear_LL/DoublyLL3.js
class Node
{
    constructor(no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLL3
{
    constructor()
    {
        console.log("Object of DoublyLL1 gets created :");
        this.first = null;
        this.iCount = 0;
    }

    InsertFirst(no)
    {
        let newn = null;

        newn = new Node(no);

        if(this.first == null)
        {
            this.first = newn;
             
        }
        else
        {
            newn.next = this.first;
            this.first.prev = newn;

            this.first = newn;
        }

        this.iCount++;
    }
    InsertLast(no)
    {
        let newn = null;
        let temp = null;

        newn = new Node(no);

        if(this.first == null)
        {
            this.first = newn;
        }
        else
        {
            temp = this.first;
            
            while(temp.next != null)
            {
                temp = temp.next;
            }
            temp.next = newn;
            newn.prev = temp;
        }

        this.iCount++;
    }
    DeleteFirst()
    {
        let temp = null;

        if(this.first == null)
        {
            return;
        }
        else if(this.first.next == null)
        {
            this.first = null;
        }
        else
        {
            temp = this.first;

            this.first = this.first.next;
        }
        this.iCount--;
    }

    DeleteLast()
    {
        let temp = null;

        if(this.first == null)
        {
            return;
        }
        else if(this.first.next == null)
        {
            this.first = null;
        }
        else
        {
            temp = this.first;

            while(temp.next.next != null)
            {
                temp = temp.next;
            }
            temp.next = null;
        }
        this.iCount--;
    }
    Display()
    {
        let temp = null;

        temp = this.first;

        while (temp != null) 
        {
            process.stdout.write("| " + temp.data + " |->");
            temp = temp.next;
        }
        console.log("null");
    }
    Count()
    {
        return this.iCount;
    }
    InsertAtPos(no,pos)
    {   
        let newn = null;
        let temp = null;

        let iCnt = 0;

        newn = new Node(no);

        newn.data = no;
        newn.next = null;

        if(pos < 1 || pos > this.iCount + 1)
        {
            console.log("Invalid Position:");
            return;
        }

        if(pos == 1)
        {
            this.InsertFirst(no);
        }
        else if(pos == this.iCount + 1)
        {
            this.InsertLast(no);
        }
        else
        {   
            temp = this.first;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;

            temp.next.prev = newn;

            temp.next = newn;

            newn.prev = temp;

            this.iCount++;
        }


    }
    DeleteAtPos(pos)
    {
        let temp = null;

        let iCnt = 0;

        if(pos < 1 || pos > this.iCount)
        {
            console.log("Invalid Position:");
            return;
        }

        if(pos == 1)
        {
            this.DeleteFirst(no);
        }
        else if(pos == this.iCount)
        {
            this.DeleteLast(no);
        }
        else
        {   
            temp = this.first;
            
            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;

            temp.next.next.prev = temp;

            this.iCount--;
        }
    }
}

function main()
{
    let obj = new DoublyLL3();

    iRet = 0;

    obj.InsertFirst(51);
    obj.InsertFirst(41);
    obj.InsertFirst(31);
    obj.InsertFirst(21);
    obj.InsertFirst(11);

    obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

    obj.InsertLast(101);
    obj.InsertLast(111);

    obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

    obj.DeleteFirst();

    obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

    obj.DeleteLast();

    obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

    obj.InsertAtPos(113, 3);

    obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

    obj.DeleteAtPos(2);

   obj.Display();

    iRet = obj.Count();

    console.log("Number of nodes are: "+iRet);

}
main();

File: DoublyLinear_LL/DoublyLL3.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLL3:
    def __init__(self):
        self.first = None
        self.iCount = 0

    def InsertFirst(self,no):

        newn = Node(no)

        if(self.first == None):
            self.first = newn

        else:
            newn.next = self.first

            self.first.prev = newn
            self.first = newn


        self.iCount = self.iCount + 1


    def InsertLast(self,no):

        newn = Node(no)

        if(self.first == None):
            self.first = newn

        else:
            
            temp = self.first

            while(temp.next != None):
                temp = temp.next


            temp.next = newn

            newn.prev = temp


        self.iCount = self.iCount + 1


    def DeleteFirst(self):
        
        temp = None

        temp = self.first

        if(self.first == None):
            return

        elif (self.first.next == None):
            self.first = None

        else:

            temp = self.first

            self.first = self.first.next

        self.iCount = self.iCount - 1

    def DeleteLast(self):
        
        temp = None

        if(self.first == None):
            return
        
        elif(self.first.next == None):
            
            self.first = None
            del self.first
        
        else:
            
            temp = self.first

            while(temp.next != None):
                temp = temp.next

            
            del temp.next
            temp.next = None

        self.iCount = self.iCount - 1

    def Display(self):
        temp = self.first

        while(temp != None):
            print("| ",temp.data," |->",end=" ")
            temp = temp.next

        print("None")

    def Count(self):
        return self.iCount

    
    def InsertAtPos(self,no,pos):
    
        newn = None

        newn = Node(no)

        newn.data = no
        newn.next = None

        if(pos < 1 or pos > self.iCount + 1):
            print("Invalid Position")
            return
        
        if(pos == 1):
            self.InsertFirst(no)

        elif(pos == self.iCount + 1):
            self.InsertLast(no)
        
        else:

            temp = self.first

            for i in range(pos - 1):
                temp = temp.next

            newn.next = temp.next

            temp.next.prev = newn

            temp.next = newn

            newn.prev = temp

            self.iCount = self.iCount + 1
    

    def DeleteAtPos(self,pos):


        if(pos < 1 or pos > self.iCount):
            print("Invalid Position")
            return

        if(pos == 1):
            self.DeleteFirst()

        elif(pos == self.iCount):
            self.DeleteLast()

        else:

            temp = self.first
            
            for i in range(pos - 1):
                temp = temp.next

            target = temp.next

            temp.next = target.next

            target.next.prev = temp

            del target

            self.iCount = self.iCount - 1


def main():

    sobj = DoublyLL3()

    sobj.InsertFirst(101)
    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    print("Elements of Linked List are :")
    sobj.Display()

    print("Number of Elements in Linekd List are :",sobj.Count())

    sobj.InsertLast(101)
    sobj.InsertLast(111)

    print("Elements of Linked List are :")
    sobj.Display()

    print("Number of Elements in Linekd List are :",sobj.Count())

    sobj.DeleteFirst()

    print("Elements of Linked List are :\n")
    sobj.Display()

    print("Number of Elements in Linekd List are :",sobj.Count())

    sobj.DeleteLast()

    print("Elements of Linked List are :\n")

    sobj.Display()

    sobj.InsertAtPos(75,4)

    print("Elements of Linked List are :\n")
    sobj.Display()

    print("Number of Elements in Linekd List are :",sobj.Count())

    sobj.DeleteAtPos(3)

    print("Elements of Linked List are :\n")
    sobj.Display()

    print("Number of Elements in Linekd List are :",sobj.Count())

if __name__ == "__main__":
    main()
